cranelift/codegen/meta/src/cdsl/instructions.rs

```diff
@@ -482,13 +485,16 @@ fn is_ctrl_typevar_candidate(
         let typ = result.type_var().unwrap();
         let free_typevar = typ.free_typevar();
 
-        // Non-polymorphic or derived from ctrl_typevar is OK.
-        if free_typevar.is_none() || &free_typevar.unwrap() == ctrl_typevar {
-            continue;
-        }
+        if let Some(tv) = typ.free_typevar() {
+            if &tv != ctrl_typevar {
+                return Err("type variable in output not derived from ctrl_typevar".into());
+            }
 
-        return Err("type variable in output not derived from ctrl_typevar".into());
+            // Variables derived from the control variable are ok, but we remember if they refine
+            // the control variable, as that still requires another type to instantiate.
+            refines_output = typ.refines_parent();
```

Comment:
I'm still wondering how this should work if the instruction has multiple value results. Right now, `refines_output` is set by the _last_ value result, which doesn't seem right. Should it be set true if _any_ value result is a refinement?
```suggestion
            if typ.refines_parent() {
                refines_output = true;
            }
```

Reply:
Yep, I think you're correct: it should be true if any result is a refinement. Thanks for catching this!